src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py

```diff
@@ -179,3 +192,27 @@ def event_bid_deleted(self, *, bid):
             except MarketException:
                 self.owner.log.exception("Error deleting InterAreaAgent offer")
         self._delete_forwarded_bid_entries(bid_info.source_bid)
+
+    def event_bid_changed(self, *, market_id, existing_bid, new_bid):
+        market = self.owner._get_market_from_market_id(market_id)
+        if market is None:
+            return
+
+        if market == self.markets.target:
+            # one of our forwarded bids was split, so save the residual bid for handling
+            # the upcoming trade event
+            assert existing_bid.id not in self.bid_trade_residual, \
```

Comment:
Sorry, I didn't get it. Why should "Bid should only change once before each trade"

Reply:
Event bid changed is emitted only once for every bid trade (and only if the bid trade is partial). If we receive twice an event that the same bid has changed on the target market, without receiving a bid trade event meanwhile, we know that something is wrong and we have received a duplicate event.